forest/shared/src/snapshot.rs

```diff
@@ -0,0 +1,569 @@
+                match blob_was_present {
+                    false => {
+                        warn!(%path, "deleting metadata without corresponding blob");
+                        let _ = std::fs::remove_file(path);
+                    }
+                    true => {
+                        let blob_path = PathBuf::from(blob_path);
+                        let metadata_path = PathBuf::from(path);
+                        blobs_and_metadata.push((blob_path, metadata_path));
+                    }
+                }
+            }
+            None => {
+                // this is the blob path
+                let metadata_path = format!("{path}{METADATA_FILE_SUFFIX}");
+                let blob_path = path;
+                let metadata_was_present = paths.remove(&metadata_path);
+                match metadata_was_present {
+                    false => {
+                        warn!(path = %blob_path, "ignored blob without corresponding metadata")
+                    }
+                    true => blobs_and_metadata.push((blob_path.into(), metadata_path.into())),
+                }
+            }
+        }
+    }
+
+    Ok(blobs_and_metadata
+        .into_iter()
+        .flat_map(|(blob, metadata)| {
+            std::fs::read_to_string(&metadata)
+                .map_err(|_| warn!(path = ?metadata, "ignoring unreadable metadata file"))
+                .and_then(|s| {
+                    serde_json::from_str(&s).map_err(
+                        |_| warn!(path = ?metadata, "ignoring invalid format for metadata file"),
+                    )
+                })
+                .map(|metadata| (blob, metadata))
+        })
+        .collect())
+}
+
+/// `file` is a snapshot file with a conventional name.
+/// This function will move `file` into `snapshot_dir`, adding an appropriate metadata
+/// file to allow it to be recognized in [list].
+pub async fn intern(
+    snapshot_dir: &Path,
+    file: &Path,
+    chain: &NetworkChain,
+    vendor: &str,
+) -> anyhow::Result<PathBuf> {
+    let (height, date, time) = file
+        .file_name()
+        .and_then(OsStr::to_str)
+        .context("non-utf8 or missing filename")
+        .and_then(parse::parse_filename)
+        .context("invalid filename")?;
+
+    intern_and_create_metadata(
+        snapshot_dir,
+        file,
+        height,
+        date,
+        time,
+        chain,
+        vendor,
+        None,
+        None,
+    )
+    .await
+}
+
+/// Metadata about a snapshot blob
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum SnapshotMetadata {
+    // We use an enum to handle forward-incompatible changes in the future
+    V1 {
+        height: i64,
+        date: NaiveDate,
+        // The `forest` vendor doesn't include time
+        time: Option<chrono::NaiveTime>,
+        chain: String,
+        vendor: String,
+        // The stable url used
+        source_url: Option<String>,
```

Comment:
Perhaps we could use a strong type here, like `Url`? Not a big deal, though.

Reply:
Closing this as we're gonna rip all this out anyway